lib/vrl/compiler/src/compiler.rs

```diff
@@ -206,21 +206,66 @@ impl<'a> Compiler<'a> {
         use ast::Assignment::*;
 
         self.state.snapshot();
+        let (_span, assignment) = node.take();
 
-        let node = node.map(|assignment| match assignment {
-            Single { target, expr } => {
+        let node = match assignment {
+            Single { target, op, expr } => {
                 let span = expr.span();
-                let expr = Box::new(expr.map(|node| self.compile_expr(Node::new(span, node))));
 
-                Variant::Single { target, expr }
+                match op {
+                    AssignmentOp::Assign => {
+                        let expr =
+                            Box::new(expr.map(|node| self.compile_expr(Node::new(span, node))));
+
+                        Node::new(span, Variant::Single { target, expr })
+                    }
+                    AssignmentOp::Merge => {
+                        let expr = Box::new(Node::new(
+                            span,
+                            Expr::Op(self.compile_op(Node::new(
+                                span,
+                                ast::Op(
+                                    Box::new(target.clone().map(|target| target.into_expr(&span))),
+                                    Node::new(span, ast::Opcode::Union),
+                                    expr,
+                                ),
+                            ))),
+                        ));
+
+                        Node::new(span, Variant::Single { target, expr })
+                    }
+                }
             }
-            Infallible { ok, err, expr } => {
+            Infallible { ok, err, op, expr } => {
                 let span = expr.span();
-                let expr = Box::new(expr.map(|node| self.compile_expr(Node::new(span, node))));
 
-                Variant::Infallible { ok, err, expr }
+                match op {
+                    AssignmentOp::Assign => {
+                        let expr =
+                            Box::new(expr.map(|node| self.compile_expr(Node::new(span, node))));
+
+                        let node = Variant::Infallible { ok, err, expr };
+                        Node::new(span, node)
+                    }
+                    AssignmentOp::Merge => {
+                        let expr = Box::new(Node::new(
+                            span,
+                            Expr::Op(self.compile_op(Node::new(
+                                span,
+                                ast::Op(
+                                    Box::new(ok.clone().map(|target| target.into_expr(&span))),
+                                    Node::new(span, ast::Opcode::Union),
+                                    expr,
+                                ),
+                            ))),
+                        ));
+
+                        let node = Variant::Infallible { ok, err, expr };
+                        Node::new(span, node)
+                    }
+                }
```

Comment:
This is where the syntax rewriting is happening. I'm sure theres a good chunk of code duplication that can be refactored. I wonder if it would be useful to enable more library support for syntax rewriting, depending on if it would be useful elsewhere too. Perhaps, ultimately we could have a stage before the compiler that does this.

This could also be extended for `+=` etc..

Reply:
I concur, we need an [intermediate representation](https://en.wikipedia.org/wiki/Intermediate_representation) layer 😛